### ISSUE TITLE
Virtual inheritance off multiple base classes fails to compile

### ISSUE THREAD

**Issue by nilsdeppe** [UIUC-PPL/charm#2049]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/2049*

---
I have a base class

```
<code class="cpp">
template <size_t Number>
struct Base : virtual PUP::able {
  PUPable_abstract(Base);
  /* ... */
};
</code>
```

and a derived class

```
<code class="cpp">
struct Derived : Base<0>, Base<1>, Base<2> {
  /* ... */
};
</code>
```

This code fails to compile because Charm++ uses C-style casts instead of a dynamic_cast in PUPable_operator_inside and PUPable_operator_outside. See:

https://github.com/UIUC-PPL/charm/blob/d994dfc2febf416f0814a5cfb23ac95cad861677/src/util/pup.h#L722

The corrected code is:

```
<code class="cpp">
#define PUPable_operator_inside(className)\
    friend inline void operator|(PUP::er &p,className &a) {a.pup(p);}\
    friend inline void operator|(PUP::er &p,className* &a) {\
	PUP::able *pa=a;  p(&pa);  a=dynamic_cast<className *>(pa);\
    }

//  Macros to be used outside a class body.
#define PUPable_operator_outside(className)\
    inline void operator|(PUP::er &p,className &a) {a.pup(p);}\
    inline void operator|(PUP::er &p,className* &a) {\
	PUP::able *pa=a;  p(&pa);  a=dynamic_cast<className *>(pa);\
    }
</code>
```

**Comment by nilsdeppe** [comment]
*Original date: 2019-01-27 00:21:08*

---
I'm hoping this could make it into 6.9.1 since it's a very small fix :)

**Comment by stwhite91** [comment]
*Original date: 2019-01-27 20:15:26*

---
We should be able to have the fix in 6.9.1, thanks

**Comment by stwhite91** [comment]
*Original date: 2019-01-28 18:44:26*

---
~~https://charm.cs.illinois.edu/gerrit/#/c/charm/+/4926/~~ https://github.com/UIUC-PPL/charm/commit/7e3230581f9313d069f8a470ce6e1f03572b9afd
